mithril-aggregator/src/artifact_builder/cardano_database_artifacts/digest.rs

```diff
@@ -1,78 +1,111 @@
+    digests_dir: PathBuf,
 
-        Ok(Self { server_url_prefix })
-    }
-}
+    immutable_file_digest_mapper: Arc<dyn ImmutableFileDigestMapper>,
 
-#[async_trait]
-impl DigestFileUploader for AggregatorDigestFileUploader {
-    async fn upload(&self) -> StdResult<DigestLocation> {
-        Ok(DigestLocation::Aggregator {
-            uri: self
-                .server_url_prefix
-                .join("artifact/cardano-database/digests")?
-                .to_string(),
-        })
-    }
-}
-
-pub struct DigestArtifactBuilder {
-    uploaders: Vec<Arc<dyn DigestFileUploader>>,
     logger: Logger,
 }
 
 impl DigestArtifactBuilder {
     /// Creates a new [DigestArtifactBuilder].
-    pub fn new(uploaders: Vec<Arc<dyn DigestFileUploader>>, logger: Logger) -> StdResult<Self> {
-        if uploaders.is_empty() {
-            return Err(anyhow!(
-                "At least one uploader is required to create an 'DigestArtifactBuilder'"
-            ));
-        }
-
+    pub fn new(
+        aggregator_url_prefix: Url,
+        uploaders: Vec<Arc<dyn DigestFileUploader>>,
+        digests_dir: PathBuf,
+        immutable_file_digest_mapper: Arc<dyn ImmutableFileDigestMapper>,
+        logger: Logger,
+    ) -> StdResult<Self> {
         Ok(Self {
+            aggregator_url_prefix,
             uploaders,
+            digests_dir,
+            immutable_file_digest_mapper,
             logger: logger.new_with_component_name::<Self>(),
         })
     }
 
     pub async fn upload(&self) -> StdResult<Vec<DigestLocation>> {
-        self.upload_digest_archive().await
+        let digest_path = self.create_digest_archive().await?;
+
+        let locations = self.upload_digest_archive(&digest_path).await;
+        fs::remove_file(&digest_path).with_context(|| {
+            format!(
+                "Could not remove digest archive file: '{}'",
+                digest_path.display()
+            )
+        })?;
+        locations
+    }
+
+    async fn create_digest_archive(&self) -> StdResult<PathBuf> {
+        let immutable_file_digest_map = self
+            .immutable_file_digest_mapper
+            .get_immutable_file_digest_map()
+            .await?
+            .into_iter()
+            .map(
+                |(immutable_file_name, digest)| CardanoDatabaseDigestListItemMessage {
+                    immutable_file_name,
+                    digest,
+                },
+            )
+            .collect::<Vec<_>>();
+
+        let digests_file_path = DigestArtifactBuilder::get_digests_file_path(&self.digests_dir);
+
+        if let Some(digests_dir) = digests_file_path.parent() {
+            fs::create_dir_all(digests_dir).with_context(|| {
+                format!(
+                    "Can not create digests directory: '{}'",
+                    digests_dir.display()
+                )
+            })?;
+        }
+
+        let digest_file = fs::File::create(digests_file_path.clone()).unwrap();
```

Comment:
This unwrap should be replaced by a returned error:
```suggestion
        let digest_file = fs::File::create(digests_file_path.clone())?;
```

Reply:
Done